vllm/v1/worker/gpu_model_runner.py

```diff
@@ -867,6 +868,9 @@ def execute_model(
         intermediate_tensors: Optional[IntermediateTensors] = None,
     ) -> Union[ModelRunnerOutput, torch.Tensor]:
         self._update_states(scheduler_output)
+        if not scheduler_output.total_num_scheduled_tokens:
+            # Return empty ModelRunnerOuptut if there's no work to do.
+            return EMPTY_MODEL_RUNNER_OUTPUT
```

Comment:
Just wondering: why not return `None`?

Reply:
Also, just curious: is this compatible with PP? cc @comaniac

Reply:
> Just wondering: why not return `None`?

I could do but I think it would actually be more intrusive, i.e. change the `execute_model` signature and require additional check in the `step()` method. This in some way is more natural since it's just "empty" output which will still get passed to `self.scheduler.update_from_output()` which is fine (it will do nothing apart from return the EngineCoreOutputs with empty stats).

Reply:
I admit I haven't fully digested the PP impl ... so yes would be great if @comaniac could comment on this!

Reply:
It seems compatible but may be further cleaned up. In short, when PP>1, I guarantee only the batches with `scheduler_output.total_num_scheduled_tokens` will be executed: https://github.com/vllm-project/vllm/blob/main/vllm/v1/engine/core.py#L181.

Specifically since `.schedule()` may schedule nothing when PP>1 (because all requests are on the fly), we already have the case that `scheduler_output.total_num_scheduled_tokens` may be 0.